### PR TITLE
From associatedtype to Any

### PR DESCRIPTION
**Pull request by dreymonde**
The problem with `associatedtype` is that we cannot create `EmailValidator` which validates a `String`, `LengthValidator` which also validates a `String` and put them into array of some abstract String validators, so `Validator` doesn't make any sense.

### CODE REVIEW

**Comment by paulofaria** [comment]
One would have to `if let cast` and throw if it fails? Something like that?

**Comment by dreymonde** [comment]
@paulofaria yes, exactly. It will even become more flexible, because then validators can work with multiple types. And I also added default `ValidationError.incompatibleType`.

**Comment by paulofaria** [comment]
LGMT. can you give me an example of a validator and its usage, please?

**Comment by dreymonde** [comment]
@paulofaria I'm working on integrating `Mapper` with `Validator` (that's how I found this problem in the first place :smile: ). So the `map(from:)` now looks like this:

``` swift
public func map<T>(from key: String, validators: [Validator] = []) throws -> T {
    let value: T = try structuredData.get(key)
    try validators.forEach { try $0.validate(value) }
    return value
}
```

And so it is possible to call:

``` swift
let value: String = mapper.map(from: "value", validators: [EmailValidator(), LengthValidator()])
```

File: Sources/Validator.swift [hunk 1]
```diff
@@ -16,4 +15,11 @@ extension Validator {
 
 public enum ValidationError: ErrorProtocol {
     case validationFailed
+    case incompatibleType
+}
+
+extension Sequence where Iterator.Element == Validator {
+    public func validate(_ value: Any) throws {
+        try self.forEach({ try $0.validate(value) })
+    }
```
Comment (dreymonde) [inline Sources/Validator.swift:48]: This one allows to write `try validators.validate(value)`

**Comment by paulofaria** [comment]
nice! I like it! very simple and efficient! (:

**Comment by paulofaria** [comment]
Oh, can you add MIT header, pleasse?

**Comment by dreymonde** [comment]
Done! Good to merge? @paulofaria

**Comment by paulofaria** [comment]
LGTM. :shipit:

**Comment by davidask** [comment]
I'm a bit late to the party, but maybe `associatedtype` isn't a bad idea. Using `Any` seems to be wasting inferred types and the use of generics, don't you think? For instance, i don't think i should be able to validate a string with integer validators. The way to go here would be to create more protocols, and extend the primitive types with them i think. Another point is that i think a validator should be able to validate optionality, by its inferred type.

I like the throwing mechanism that can be ignored using `isValid`!

**Comment by dreymonde** [comment]
Of course you shouldn’t validate a string with integer validator, in that case you just `throw ValidationError.incompatibleType`. The problem with `associatedtype` is, one more time, we can’t create an array of «string validators», that’s just impossible. All we can do is to create an array of validators of the same type (same type of validators themselves), e.g. `Array<EmailValidator>`, which makes almost no sense.

However, one can create something like that if he wants:

``` swift
protocol StringValidator: Validator {
    func validate(string: String) throws
}

extension StringValidator {
    func validate(_ value: Any) throws {
        guard let string = value as? String else {
            throw ValidationError.incompatibleType
        }
        try validate(string: string)
    }
}

// Still Validator
struct LengthValidator: StringValidator {
    let length: Int
    func validate(string: String) throws {
        if string.characters.count < length {
            throw Error.invalidLength
        }
    }

    enum Error: ErrorProtocol {
        case invalidLength
    }
}
```

**Comment by davidask** [comment]
Why not this?

``` swift
public protocol Validator {
    associatedtype Validating
    associatedtype Error: ErrorProtocol
    func validate(_ value: Validating) throws
}

extension Validator {
    public func isValid(_ value: Validating) -> Bool {
        do {
            try validate(value)
            return true
        } catch {
            return false
        }
    }
}

extension Sequence where Iterator.Element: Validator {
    public func validate(_ value: Iterator.Element.Validating) throws {
        try self.forEach {
            try $0.validate(value)
        }
    }
}


func testValueWithValidators<T: Any, V: Validator where V.Validating == T>(value: T, validators: [V]) throws {
    try validators.validate(value)
}


enum StringValidator: Validator {
    enum Error: ErrorProtocol {
        case LengthError(given: Int, allowed: Range<Int>)
    }

    case Length(Range<Int>)

    func validate(_ value: String) throws {
        switch self {
        case .Length(let range):
            guard range.contains(value.characters.count) else {
                throw Error.LengthError(given: value.characters.count, allowed: range)
            }
            break
        }
    }
}


do {
    try testValueWithValidators(value: "HELLO MY NAME IS DAVID", validators: [StringValidator.Length(0..<3)])
}
catch {
    print("\(error)")
}
```

**Comment by davidask** [comment]
How about i create a PR? Next step would be inferred validation for optionals.

**Comment by dreymonde** [comment]
@formbound that wouldn’t work. If you create another `Validator` which works with string, for example, `EmailValidator`, you won’t be able to do `[StringValidator.Length(0..<3), EmailValidator()]`. And I see your approach with enums, but that’s not extendable — you can’t add new cases in the extensions.

**Comment by davidask** [comment]
@dreymonde I see what you mean. My point here is i think the solution with `Any` is a bit Java, and it's basically just more boilerplate code that doesn't do anything. In my example, an e-mail validator would be part of a string validator that we'd supply – but yes extending when needed would be a problem.

Another, still better, way to go would be to do something like this. Let people create their own custom validators without having to create yet another type, and always check for type.

``` swift
public protocol Validator {
    associatedtype Validating
    associatedtype Error: ErrorProtocol
    var block: Validating throws -> Void { get }
    init(block: Validating throws -> Void)
}

extension Validator {

    public func validate(_ value: Validating) throws {
        try block(value)
    }

    public func isValid(_ value: Validating) -> Bool {
        do {
            try validate(value)
            return true
        } catch {
            return false
        }
    }
}

extension Sequence where Iterator.Element: Validator {
    public func validate(_ value: Iterator.Element.Validating) throws {
        try self.forEach {
            try $0.validate(value)
        }
    }
}

func testValueWithValidators<T: Any, V: Validator where V.Validating == T>(value: T, validators: [V]) throws {
    try validators.validate(value)
}

struct StringValidator: Validator {
    struct Error: ErrorProtocol {
        let description: String
    }

    let block: String throws -> Void

    init(block: String throws -> Void) {
        self.block = block
    }

    static let emailValidator = StringValidator {
        value in

        guard value.characters.contains("@") else {
            throw Error(description: "No @-sign in e-mail")
        }
    }

    static func lengthValidator(range: Range<Int>) -> StringValidator {
        return StringValidator {
            value in

            guard range.contains(value.characters.count) else {
               throw Error(description: "Not long or short enough")
            }
        }
    }
}

do {
    try testValueWithValidators(value: "String", validators: [StringValidator.emailValidator, StringValidator.lengthValidator(range: 0..<3)])
}
catch {
    print("\(error)")
}
```

**Comment by paulofaria** [comment]
`Any` would allow a validator to support more than one type with the same logic.

**Comment by paulofaria** [comment]
I think the whole issue comes to… can we make everything we want using validators of the same type? @dreymonde do you have a "real life" example which needs validators with different types?

**Pull request merged**